crates/consistent-hashing/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+            bits: builder.bit_mask() & !((n + 2).next_power_of_two() as u64 / 2 - 1),
+            stack: if n == 0 { vec![0] } else { vec![] },
+            builder,
+            n,
+        }
+    }
+}
+
+impl<H: HashSeqBuilder> Iterator for ConsistentHashIterator<H> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(res) = self.stack.pop() {
+            return Some(res);
+        }
+        while self.bits > 0 {
+            let bit = self.bits & !(self.bits - 1);
+            self.bits &= self.bits - 1;
+            let inner = BucketIterator::new(bit as usize * 2, bit, self.builder.hash_seq(bit));
+            self.stack = inner.take_while(|x| *x >= self.n).collect();
+            if let Some(res) = self.stack.pop() {
+                return Some(res);
+            }
+        }
+        None
+    }
+}
+
+/// Wrapper around `ConsistentHashIterator` and `ConsistentHashRevIterator` to compute
+/// the next or previous consistent hash for a given key for a given number of nodes `n`.
+pub struct ConsistentHasher<H: HashSeqBuilder> {
+    builder: H,
+}
+
+impl<H: HashSeqBuilder + Clone> ConsistentHasher<H> {
+    pub fn new(builder: H) -> Self {
+        Self { builder }
+    }
+
+    pub fn prev(&self, n: usize) -> Option<usize> {
+        let mut sampler = ConsistentHashRevIterator::new(n, self.builder.clone());
+        sampler.next()
+    }
+
+    pub fn next(&self, n: usize) -> Option<usize> {
+        let mut sampler = ConsistentHashIterator::new(n, self.builder.clone());
+        sampler.next()
+    }
+}
+
+/// Implementation of a consistent choose k hashing algorithm.
+/// It returns k distinct consistent hashes in the range `0..n`.
+/// The hashes are consistent when `n` changes and when `k` changes!
+/// I.e. on average exactly `1/(n+1)` (resp. `1/(k+1)`) many hashes will change
+/// when `n` (resp. `k`) increases by one. Additionally, the returned `k` tuple
+/// is guaranteed to be uniformely chosen from all possible `n-choose-k` tuples.
+pub struct ConsistentChooseKHasher<H: ManySeqBuilder> {
+    builder: H,
+    k: usize,
+}
+
+impl<H: ManySeqBuilder> ConsistentChooseKHasher<H> {
+    pub fn new(builder: H, k: usize) -> Self {
+        Self { builder, k }
+    }
+
+    // TODO: Implement this as an iterator!
+    pub fn prev(&self, mut n: usize) -> Vec<usize> {
+        let mut samples = Vec::with_capacity(self.k);
+        let mut samplers: Vec<_> = (0..self.k)
+            .map(|i| ConsistentHashRevIterator::new(n - i, self.builder.seq_builder(i)).peekable())
+            .collect();
+        for i in (0..self.k).rev() {
+            let mut max = 0;
+            for k in 0..=i {
+                while samplers[k].peek() >= Some(&(n - k)) && n - k > 0 {
+                    samplers[k].next();
+                }
+                max = max.max(samplers[k].peek().unwrap() + k);
+            }
+            samples.push(max);
+            n = max;
+        }
+        samples.sort();
```

Comment:
Is it important that we sort this list within our function? If the user wants the list sorted they could do it themselves?